### PR TITLE
removed direct Workspace access from WorkspaceAnalyzerOptions and made OOP analyzers to use snapshot of OptionSet

### PR DESCRIPTION
**Pull request by heejaechang**
**Customer scenario**

Customer has either code lens/FSA on, and just opened new solution. and in some cases, VS sync some data twice.

**Bugs this fixes:**

This doesn't have an associate bug, but one of PRs to improve OOP perf on the initial solution synchronization.

**Workarounds, if any**

There is no workaround. before the very first primary solution synchronization, all solution treated as temporary solution and not reused to incrementally update the solution.

**Risk**

Low. There should be no functional behavior change.

**Performance impact**

Should make initial solution perf better in OOP (not this PR alone but with next PR combined - https://github.com/dotnet/roslyn/pull/19311)

**Is this a regression from a previous update?**

No

**Root cause analysis:**

This is not regression. it was just the way it is designed. now I am trying to make this particular case to have better perf by not making all solution a temporary solution before the very first primary workspace synchronization.

**How was the bug found?**

While doing FAR/GoTo OOP work.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/Document.cs [hunk 1]
```diff
@@ -464,13 +464,22 @@ private string GetDebuggerDisplay()
         /// </remarks>
         public Task<DocumentOptionSet> GetOptionsAsync(CancellationToken cancellationToken = default(CancellationToken))
         {
+            return GetOptionsAsync(Project.Solution.Options, cancellationToken);
+        }
+
+        internal Task<DocumentOptionSet> GetOptionsAsync(OptionSet globalOptionSet, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Document.cs]: No optoinal cancellation tokens on non-public-api methods

**Comment by CyrusNajmabadi** [comment]
Tagging @jasonmalinowski

File: src/Workspaces/Core/Portable/Workspace/Solution/Document.cs [hunk 2]
```diff
@@ -464,13 +464,22 @@ private string GetDebuggerDisplay()
...
+        {
+            // TODO: we have this workaround since Solution.Options is not actually snapshot but just return Workspace.Options which violate snapshot model.
+            //       this doesn't validate whether same optionset is given to invalidate the cache or not. this is not new since existing implementation
+            //       also didn't check whether Workspace.Option is same as before or not. all wierd-ness come from the root cause of Solution.Options violating
+            //       snapshot model. once that is fixed, we can remove this workaround
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/Document.cs]: How much work is it to just fix the weirdness? This was intended to have been cleaned up by 2017 RTM but didn't happen, obviously.

Comment (heejaechang) [inline src/Workspaces/Core/Portable/Workspace/Solution/Document.cs]: opened an issue. https://github.com/dotnet/roslyn/issues/19284

not sure exactly how much work. but probably some work.

**Comment by heejaechang** [comment]
@CyrusNajmabadi are  you going to sign off? @jinujoseph 

are you going to push FAR/GOTO on 15.3? does this need to go in master?

File: src/Features/Core/Portable/Diagnostics/WorkspaceAnalyzerOptions.cs [hunk 3]
```diff
@@ -10,13 +14,35 @@ namespace Microsoft.CodeAnalysis.Diagnostics
...
+            }
+
+            var document = _workspace.CurrentSolution.GetDocument(documentId);
+            if (document == null)
+            {
+                return _optionSet;
+            }
+
+            var documentOptionSet = await document.GetOptionsAsync(_optionSet, cancellationToken).ConfigureAwait(false);
+            return documentOptionSet ?? _optionSet;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Diagnostics/WorkspaceAnalyzerOptions.cs]: can document.GetOptionsAsync actuall return null?  @jasonmalinowski

File: src/Features/Core/Portable/Diagnostics/WorkspaceAnalyzerOptions.cs [hunk 4]
```diff
@@ -10,13 +14,35 @@ namespace Microsoft.CodeAnalysis.Diagnostics
...
+            _workspace = workspace;
+            _optionSet = optionSet;
         }
 
-        public Workspace Workspace { get; }
+        public HostWorkspaceServices Services => _workspace.Services;
+
+        public async Task<OptionSet> GetDocumentOptionSetAsync(SyntaxTree syntaxTree, CancellationToken cancellationToken)
+        {
+            var documentId = _workspace.CurrentSolution.GetDocumentId(syntaxTree);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Diagnostics/WorkspaceAnalyzerOptions.cs]: i don't like this.  why are we checking against CurrentSolution?   This seems like a race-nightmare.  Why not just grab the Solution snapshot when we make this object.  That way it is always valid as we're doing analysis, and we can always use it to map from a SyntaxTree to a document it?

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.CompilationManager.cs [hunk 5]
```diff
@@ -118,7 +118,7 @@ public Task<CompilationWithAnalyzers> CreateAnalyzerDriverAsync(Project project,
                 // in IDE, we always set concurrentAnalysis == false otherwise, we can get into thread starvation due to
                 // async being used with syncronous blocking concurrency.
                 return new CompilationWithAnalyzersOptions(
-                    options: new WorkspaceAnalyzerOptions(project.AnalyzerOptions, project.Solution.Workspace),
+                    options: new WorkspaceAnalyzerOptions(project.AnalyzerOptions, project.Solution.Options, project.Solution.Workspace),
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.CompilationManager.cs]: why not grab project.Solution instead in here.  It holds onto a Solutoin snapshot.  but that seems appropriate as we want to be doing all analysis and getting all options in the context of that snapshot.

File: src/Workspaces/Core/Portable/Workspace/Solution/Document.cs [hunk 6]
```diff
@@ -464,13 +464,22 @@ private string GetDebuggerDisplay()
         /// </remarks>
         public Task<DocumentOptionSet> GetOptionsAsync(CancellationToken cancellationToken = default(CancellationToken))
         {
+            return GetOptionsAsync(Project.Solution.Options, cancellationToken);
+        }
+
+        internal Task<DocumentOptionSet> GetOptionsAsync(OptionSet globalOptionSet, CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Document.cs]: solutionOptions

Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Document.cs]: Why did you need a helper method?  Who else will call this?

File: src/Features/Core/Portable/Diagnostics/WorkspaceAnalyzerOptions.cs (hunk 3 above)
Comment (jasonmalinowski) [inline src/Features/Core/Portable/Diagnostics/WorkspaceAnalyzerOptions.cs]: No.

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.CompilationManager.cs (hunk 5 above)
Comment (heejaechang) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.CompilationManager.cs]: are you asking me or @jasonmalinowski ? I didn't change this semantic. also, I am not sure whether in current design where analyzer lives in compiler layer there is a way to pass in solution along with analyzer bits to compiler.

Comment (heejaechang) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.CompilationManager.cs]: ah, actually, yes. let me change it to that. I was focusing on keeping the semantic. that change seems better.

File: src/Features/Core/Portable/Diagnostics/WorkspaceAnalyzerOptions.cs (hunk 4 above)
Comment (heejaechang) [inline src/Features/Core/Portable/Diagnostics/WorkspaceAnalyzerOptions.cs]: yep.

File: src/Features/Core/Portable/Diagnostics/WorkspaceAnalyzerOptions.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Diagnostics/WorkspaceAnalyzerOptions.cs]: remove ?? then @heejaechang :)

File: src/Features/Core/Portable/Diagnostics/WorkspaceAnalyzerOptions.cs (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Diagnostics/WorkspaceAnalyzerOptions.cs]: So my concern is this.  You're passing a Workspace + OptionSet, but then using Workspace.CurrentSolution instead of the actual Solution that we started all the work against.  That just seems busted (regardless of if this is what we already do).  I would just grab hte Solution snapshot.

File: src/Workspaces/Core/Portable/Workspace/Solution/Document.cs (hunk 6 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Workspace/Solution/Document.cs]: this is called in OOP side to pass in option set.

File: src/Features/Core/Portable/Diagnostics/WorkspaceAnalyzerOptions.cs (hunk 3 above)
Comment (heejaechang) [inline src/Features/Core/Portable/Diagnostics/WorkspaceAnalyzerOptions.cs]: sure. but the original code was with ??
..
var documentOptionSet = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);		 +            return workspaceAnalyzerOptions.GetDocumentOptionSetAsync(syntaxTree, cancellationToken);
 -            return documentOptionSet ?? workspace.Options;
..

**Comment by heejaechang** [comment]
@MattGertz can I get sign off for checkin for master?

**Comment by heejaechang** [comment]
retest windows_debug_unit64_prtest please

**Comment by heejaechang** [comment]
retest windows_release_vs-integration_prtest please
